app/code/Magento/Catalog/Block/Product/ProductList/Toolbar.php

```diff
@@ -192,7 +192,14 @@ public function setCollection($collection)
             $this->_collection->setPageSize($limit);
         }
         if ($this->getCurrentOrder()) {
-            $this->_collection->setOrder($this->getCurrentOrder(), $this->getCurrentDirection());
+            if (($this->getCurrentOrder()) == 'position') {
+                $this->_collection->addAttributeToSort(
+                    $this->getCurrentOrder(),
+                    $this->getCurrentDirection()
+                )->addAttributeToSort('entity_id', $this->getCurrentDirection());
```

Comment:
Assuming that this collection contains products – using `entity_id` directly is discouraged, as this may break on certain scenarios in Magento Commerce (formerly EE) with staging feature enabled.
This may be resolved using metadata for the entity from `MetadataPool` and getting proper link field for the entity in question. You may search code for such examples (something I've found right away: `\Magento\ConfigurableProduct\Model\ResourceModel\Indexer\Stock\Configurable::_getStockStatusSelect`.

Reply:
Then you suggest us a proper solutions for this issue. Please let me know what exactly I need to change in above code.